### PR TITLE
[46] Media download

### PR DESCRIPTION
**Pull request by lichingt**
[46] Media download

Enable download of mp4, gif, png

Resolves planimation/frontend#46

### CODE REVIEW

File: server/app/media_export/Media_exporter.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import gif_exporter
+
+def export_media(vfg_file, format, parameters):
+
+    if parameters==None:
+        raise Exception("Parameters not found")
+
+    # Extract parameters
+    startStep = int(parameters.get("startStep", 0))
+    stopStep  = int(parameters.get("stopStep", 999999))
```
Comment (nirlipo) [inline server/app/media_export/Media_exporter.py]: Not a dealbraker, but consider using a defined constant, e.g. MAX_PLAN_LENGTH or use the system sys.maxsize

Comment (nirlipo) [reply server/app/media_export/Media_exporter.py]: You can define the constants similar to how you defined below the gif parameters

File: server/app/media_export/gif_exporter.py [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        try:
+            image_data = base64.b64decode(base64_str)
+            image = Image.open(BytesIO(image_data))
+            image = image.convert("RGBA")
+            image_table[key] = np.array(image)
+        except:
+            print(f"Warning: Skipping '{key}' because its base64 value could not be decoded.")
+            continue
+
+    def update(frame):
```
Comment (nirlipo) [inline server/app/media_export/gif_exporter.py]: Do you need to define this function inside the create_GIF function?

File: server/app/media_export/mp4_exporter.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        try:
+            image_data = base64.b64decode(base64_str)
+            image = Image.open(BytesIO(image_data))
+            image = image.convert("RGBA")
+            image_table[key] = np.array(image)
+        except:
+            print(f"Warning: Skipping '{key}' because its base64 value could not be decoded.")
+            continue
+
+    def update(frame):
```
Comment (nirlipo) [inline server/app/media_export/mp4_exporter.py]: same comment as above.

File: server/app/media_export/png_exporter.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+def apply_tint(image, color):
+    r, g, b, a = np.rollaxis(image, axis=-1)
+    r = np.clip(np.uint8(r * color[0]), 0, 255)
+    g = np.clip(np.uint8(g * color[1]), 0, 255)
+    b = np.clip(np.uint8(b * color[2]), 0, 255)
+    tinted_image = np.stack([r, g, b, a], axis=-1)
+    return tinted_image
+
+def create_PNGs(vfg_file, startStep=_START_DEFAULT, stopStep=_END_DEFAULT):
```
Comment (nirlipo) [inline server/app/media_export/png_exporter.py]: Given that the mp4, gif and png functions are quite similar, would it be better to refactor it and reuse code so it is easier to maintain?

**Review by nirlipo — CHANGES_REQUESTED** [review]
Minor changes regarding the constant, and let me know what do you think regarding the comment on refactoring the code in the gif/mp4/png_exporter.py files so the common code gets reused. Thanks!

**Comment by lichingt** [comment]
Constants have been added and codes refactored. Please let me know if further updates is required. Thank you!

**Review by nirlipo — APPROVED** [review]
Great! Ready to be merged.

**Pull request merged**